### ISSUE TITLE
gRPC-PHP persistent channel doesn't work on php-fpm mode since v1.9.0

### ISSUE THREAD

**Issue by ZhouyihaiDing** [grpc/grpc#14713]
Split from the issue #14534

When running spanner API in php-fpm mode, the first RPC still takes long time until getting the response, which seems like the channel is not been cached. 

I am running this code under php-fpm mode with 2 php workers on my local desktop, and the spanner region located in us-central:
```
putenv("./key.json");
require('./vendor/autoload.php');
use Google\Cloud\Spanner\SpannerClient;

$spanner = new SpannerClient(['projectId' =>'']);
$database = $spanner->connect('instance', 'database');
$then = microtime(true);
$results = $database->execute('SELECT * from Singers');
foreach ($results as $data) {
    print_r($data);
}
$elapsed = (microtime(true) - $then) * 1000;
echo "First Query: ".number_format($elapsed, 3)." milliseconds".PHP_EOL;
$then = microtime(true);
$results = $database->execute('SELECT * from Singers2');
foreach ($results as $data) {
    print_r($data);
}
$elapsed = (microtime(true) - $then) * 1000;
echo "Second Query: ".number_format($elapsed, 3)." milliseconds".PHP_EOL;
$then = microtime(true);
$results = $database->execute('SELECT * from Singers3');
foreach ($results as $data) {
    print_r($data);
}
$elapsed = (microtime(true) - $then) * 1000;
echo "Third Query: ".number_format($elapsed, 3)." milliseconds".PHP_EOL;
$database->close();
```

The first RPC latency for v1.10.0 is
`1626ms, 1697ms, 1678ms, 1488ms, 1523ms, 1508ms, 1296ms`
The first RPC latency for v1.9.0 is
`1697ms, 2262ms, 1436ms, 771ms, 800ms, 1413ms, 1319ms`
The first RPC latency for v1.8.5 is
`1223ms, 1142ms, 648ms, 656ms, 640ms. 623ms, 651ms`

**Comment by ZhouyihaiDing** [comment]
The only changes in the php part are https://github.com/grpc/grpc/pull/14167 and https://github.com/grpc/grpc/pull/14149. I am going to patch this 2 PRs into v1.8.5 to check it is caused by core or php extension.

**Comment by ZhouyihaiDing** [comment]
Caused with gRPC PHP extension part. With core 1.9.0, persistent channel works well when installing php extension without 2 PRs above.

**Comment by ZhouyihaiDing** [comment]
Hi  @axot
I think [this 3 lines](https://github.com/grpc/grpc/blob/master/src/php/ext/grpc/channel.c#L87) delete the persistent channel at the end of the request, which lead to the unexpected behavior. I am trying to fix it ASAP.

In the mean time, can you please help me to double verify it in your environment by deleting those 3 lines in `src/php/ext/grpc/channel.c` and then [install the grpc from the source](https://github.com/grpc/grpc/tree/master/src/php#build-from-source)?
```
    if (p->wrapper->ref_count == 0) {
      is_last_wrapper = true;
    }
```

The command installing gRPC php extension looks like below:
```
# install pre-request:
sudo apt-get install build-essential autoconf libtool
sudo apt-get install zip unzip zlib1g-dev

# download grpc
git clone https://github.com/grpc/grpc.git
cd grpc
git checkout v1.10.0
git submodule update --init
# vim config.m4 and change CXXFLAGS="-std=c++11" to CXXFLAGS="-std=c++11 -fno-exceptions -fno-rtti"

# install protobuf
cd third_party/protobuf
./autogen.sh
./configure
make -j8
sudo make install
export LD_LIBRARY_PATH=/usr/local/lib
cd ../..

#install grpc
make -j8
sudo make install

# install grpc php extension
cd src/php/ext/grpc
phpize
./configure
# vim channel.c and remove line 86-89
make
sudo make install
```
Then if you run `php -d extension=grpc.so --re grpc | head -1`, you can see the version of grpc.

Thank you for your help!

**Comment by axot** [comment]
@ZhouyihaiDing I'm beginning to verity this patch ;D

**Comment by axot** [comment]
@ZhouyihaiDing Here is the result:

## version
```
$ php --re grpc | head -1
Extension [ <persistent> extension #57 grpc version 1.10.0 ] {
```

## without php-fpm
```
$ php public/healthcheck/index2.php
pid: 13625
GRPC Version: 1.10.0
Warmup: 0.131 milliseconds
1 Query: 230.905 milliseconds
2 Query: 38.252 milliseconds
3 Query: 3.811 milliseconds
4 Query: 9.742 milliseconds
5 Query: 3.904 milliseconds
6 Query: 3.639 milliseconds
7 Query: 3.880 milliseconds
8 Query: 3.420 milliseconds
9 Query: 3.127 milliseconds
10 Query: 3.801 milliseconds
```

## with php-fpm
```
pid: 13479
GRPC Version: 1.10.0
Warmup: 0.116 milliseconds
1 Query: 10.957 milliseconds
2 Query: 3.691 milliseconds
3 Query: 3.736 milliseconds
4 Query: 2.715 milliseconds
5 Query: 2.599 milliseconds
6 Query: 2.551 milliseconds
7 Query: 2.755 milliseconds
8 Query: 3.369 milliseconds
9 Query: 2.547 milliseconds
10 Query: 3.147 milliseconds
```

**Comment by ZhouyihaiDing** [comment]
Thanks @axot 

In the upstream issue, the first connection of v1.10.0 takes around 1s which looks strange, because this patch only prevent the channel been deleted at the end of the script. Maybe this 230ms is using the cache of something in the previews RPC, but still need to fetch something from the auth or backend?

And in the php-fpm mode, 10ms seems like it is reuse everything.

It seems removing those 3 lines can solve the persistent channel issue temporarily in your environment too. I'll focus on making a better solution for it ASAP.

**Comment by axot** [comment]
Thanks for your nice work @ZhouyihaiDing !

It's wired the patch which fixed `persistent channel` should not boost the first connection from ~1000ms to ~200ms. Without php-fpm `persistent channel` will not work with php itself right?

**Comment by axot** [comment]
Some updates here, the `1 Query` of patch vs original are 245.656ms and 500.166ms respectively, but the tcplife is almost at same level 230.41ms and 191.12ms, so I guess there are some overheads.

## patched 1.10.0
```
GRPC Version: 1.10.0
PID    COMM        LADDR              LPORT  RADDR                  RPORT  TX_KB  RX_KB  MS
32442  grpc_execu  ::ffff:10.22.0.10  37948  ::ffff:172.217.25.74   443    7      5      942.79
32491  grpc_execu  ::ffff:10.22.0.10  38584  ::ffff:172.217.25.202  443    6      5      230.41

pid: 32552
GRPC Version: 1.10.0
Warmup: 0.108 milliseconds
1 Query: 245.656 milliseconds
2 Query: 48.934 milliseconds
3 Query: 4.811 milliseconds
4 Query: 9.284 milliseconds
5 Query: 3.965 milliseconds
6 Query: 3.683 milliseconds
7 Query: 8.677 milliseconds
8 Query: 9.088 milliseconds
9 Query: 3.934 milliseconds
10 Query: 3.288 milliseconds
```

## original 1.10.0
```
PID  COMM        LADDR              LPORT  RADDR                  RPORT  TX_KB  RX_KB  MS
882  grpc_execu  ::ffff:10.22.0.10  38388  ::ffff:172.217.24.138  443    7      5      807.20
907  grpc_execu  ::ffff:10.22.0.10  38300  ::ffff:172.217.25.74   443    6      5      191.12

GRPC Version: 1.10.0
Warmup: 0.130 milliseconds
1 Query: 500.166 milliseconds
2 Query: 41.561 milliseconds
3 Query: 3.880 milliseconds
4 Query: 3.247 milliseconds
5 Query: 3.436 milliseconds
6 Query: 4.213 milliseconds
7 Query: 3.204 milliseconds
8 Query: 3.669 milliseconds
9 Query: 3.071 milliseconds
10 Query: 3.661 milliseconds
```

**Comment by axot** [comment]
@ZhouyihaiDing I think I may found what caused this difference. In `pecl-grpc`, it seems did not include `-O2` optimization FLAGS, could you confirm this, please?

**Comment by tmatsuo** [comment]
@axot Re: cache, I recommend that you use sth like sysv cache (https://github.com/GoogleCloudPlatform/google-cloud-php/issues/819#issuecomment-365461933) because it is super fast. Only caveat is that this only works on *nix systems.

The sysv cache can be shared among processes, so that you can warm up the cache at the server startup once, then all the other processes will benefit from this.

**Comment by axot** [comment]
@tmatsuo Thank you for providing a shared memory cache implementation. I'll try this in the next week.

**Comment by axot** [comment]
@ZhouyihaiDing Any progress of persistent channel management could share ;D

**Comment by ZhouyihaiDing** [comment]
@axot I have a [draft PR](https://github.com/ZhouyihaiDing/grpc/pull/13) to my own repo so far and am  working on it.

**Comment by ZhouyihaiDing** [comment]
It should be solved by #15218.
The memory usage is protected by the upper bound for each target, so that the memory won't go infinite if keeping create new channels under php-fpm mode.
By default, each target can persist one channel, which can be re-used if the next script creates a same channel/client. The upper bound can be set as `"grpc_target_persist_bound" => $bound`.
